pkg/dataobj/index/index_builder.go

```diff
@@ -0,0 +1,510 @@
+	}
+
+	// Allocate a single buffer
+	flushBuffer := bytes.NewBuffer(make([]byte, int(float64(cfg.BuilderConfig.TargetObjectSize)*1.2)))
+
+	// Set up queues to download the next object (I/O bound) while processing the current one (CPU bound) in order to maximize throughput.
+	// Setting the channel buffer sizes caps the total memory usage by only keeping up to 3 objects in memory at a time: One being processed, one fully downloaded and one being downloaded from the queue.
+	downloadQueue := make(chan metastore.ObjectWrittenEvent, cfg.EventsPerIndex)
+	downloadedObjects := make(chan downloadedObject, 1)
+
+	s := &Builder{
+		cfg:               cfg,
+		client:            eventConsumerClient,
+		logger:            logger,
+		builder:           builder,
+		bucket:            bucket,
+		flushBuffer:       flushBuffer,
+		builderMtx:        sync.Mutex{},
+		downloadedObjects: downloadedObjects,
+		downloadQueue:     downloadQueue,
+		metrics:           metrics,
+
+		bufferedEvents: make(map[string][]metastore.ObjectWrittenEvent),
+	}
+
+	s.Service = services.NewBasicService(nil, s.run, s.stopping)
+
+	return s, nil
+}
+
+func (p *Builder) run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	p.ctx = ctx
+	p.cancel = cancel
+
+	p.wg.Add(1)
+	go func() {
+		// Download worker
+		defer p.wg.Done()
+		for event := range p.downloadQueue {
+			objLogger := log.With(p.logger, "object_path", event.ObjectPath)
+			downloadStart := time.Now()
+
+			objectReader, err := p.bucket.Get(p.ctx, event.ObjectPath)
+			if err != nil {
+				p.downloadedObjects <- downloadedObject{
+					event: event,
+					err:   fmt.Errorf("failed to fetch object from storage: %w", err),
+				}
+				continue
+			}
+
+			object, err := io.ReadAll(objectReader)
+			_ = objectReader.Close()
+			if err != nil {
+				p.downloadedObjects <- downloadedObject{
+					event: event,
+					err:   fmt.Errorf("failed to read object: %w", err),
+				}
+				continue
+			}
+			level.Info(objLogger).Log("msg", "downloaded object", "duration", time.Since(downloadStart), "size_mb", float64(len(object))/1024/1024, "avg_speed_mbps", float64(len(object))/time.Since(downloadStart).Seconds()/1024/1024)
+			p.downloadedObjects <- downloadedObject{
+				event:       event,
+				objectBytes: &object,
+			}
+		}
+	}()
+
+	level.Info(p.logger).Log("msg", "started index builder service")
+	for {
+		fetches := p.client.PollRecords(ctx, -1)
+		if fetches.IsClientClosed() || ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if errs := fetches.Errors(); len(errs) > 0 {
+			level.Error(p.logger).Log("msg", "error fetching records", "err", errs)
+			continue
+		}
+		if fetches.Empty() {
+			continue
+		}
+		fetches.EachPartition(func(ftp kgo.FetchTopicPartition) {
+			// TODO(benclive): Verify if we need to return re-poll ASAP or if sequential processing is good enough.
+			for _, record := range ftp.Records {
+				p.processRecord(record)
+			}
+		})
```

Comment:
What is the partition key for the object written events here? Considering that `p.processRecord` is per event a `p.buildIndex` which is time-consuming (right? more than a couple milliseconds) then you don't benefit much from the having the event records being partitioned. The next  `client.PollRecords` is blocked as long as `p.processRecords` is executing in the for loop. This leaves room for improvement and is certainly a day 2 optimization to parallize building the indices per partition if possible. I guess this is not solved now because you need to define which partition key is best for event object, or do I miss something?

Reply:
The relevant Produce happens at `pkg/dataobj/consumer/partition_processor.go:224`
I don't explicitly set a partition key but I was thinking to make it a tenant ID.
As long as a single builder can keep up with the throughput of a single tenant, which I think is likely, this won't be an issue.
I'm holding off on this decision until I know whether we will be handling multi-tenant logs objects or not, however.